libs/vkd3d/resource.c

```diff
@@ -1727,6 +1734,98 @@ HRESULT vkd3d_sampler_state_create_static_sampler(struct vkd3d_sampler_state *st
     return S_OK;
 }
 
+static VkResult vkd3d_sampler_state_create_descriptor_pool(struct d3d12_device *device, VkDescriptorPool *vk_pool)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkDescriptorPoolCreateInfo pool_info;
+    VkDescriptorPoolSize pool_size;
+
+    pool_size.type = VK_DESCRIPTOR_TYPE_SAMPLER;
+    pool_size.descriptorCount = 16384;
+
+    pool_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_POOL_CREATE_INFO;
+    pool_info.pNext = NULL;
+    pool_info.flags = VK_DESCRIPTOR_POOL_CREATE_FREE_DESCRIPTOR_SET_BIT;
+    pool_info.maxSets = 4096;
+    pool_info.poolSizeCount = 1;
+    pool_info.pPoolSizes = &pool_size;
+
+    return VK_CALL(vkCreateDescriptorPool(device->vk_device, &pool_info, NULL, vk_pool));
+}
+
+HRESULT vkd3d_sampler_state_allocate_descriptor_set(struct vkd3d_sampler_state *state,
+        struct d3d12_device *device, VkDescriptorSetLayout vk_layout, VkDescriptorSet *vk_set,
+        VkDescriptorPool *vk_pool)
+{
+    const struct vkd3d_vk_device_procs *vk_procs = &device->vk_procs;
+    VkResult vr = VK_ERROR_OUT_OF_POOL_MEMORY;
+    VkDescriptorSetAllocateInfo alloc_info;
+    size_t i;
+    int rc;
+
+    if ((rc = pthread_mutex_lock(&state->mutex)))
+    {
+        ERR("Failed to lock mutex, rc %d.\n", rc);
+        return hresult_from_errno(rc);
+    }
+
+    alloc_info.sType = VK_STRUCTURE_TYPE_DESCRIPTOR_SET_ALLOCATE_INFO;
+    alloc_info.pNext = NULL;
+    alloc_info.descriptorSetCount = 1;
+    alloc_info.pSetLayouts = &vk_layout;
+
+    for (i = 1; i <= state->vk_descriptor_pool_count; i++)
```

Comment:
This loop is kinda confusing. I don't understand why it starts at 1 and uses <=.

Reply:
Basically the assumption is that the most recently created pool is the easiest one to allocate a set from, so we iterate in reverse order. I can add a comment.

Reply:
I figured it was accessing the array in reverse after staring at it for a bit longer, would be easier to understand if i is decrementing i, or a comment.

Reply:
Actually, since this isn't perf-critical anyway, I'll just change it to iterate normally.